Add readme and license (#3)
esa-prl/platform_driver_ethercat_ros2: README.md

# Package Name

## Overview

> This package provides a ROS 2 lifecycle node to interface with the [platform_driver_ethercat] library for accessing Elmo drives and ATI force torque sensors via EtherCAT.

Reviewer: ```suggestion
This package provides a ROS2 lifecycle node to interface with the [platform_driver_ethercat] library for accessing Elmo drives and ATI force torque sensors via EtherCAT.
```

Author: Consensus was reached to stick to the official spelling which includes a space

**Keywords:** ati fts, elmo, ethercat, joint control, soem

### License

The source code is released under a [GPLv3 License](https://www.gnu.org/licenses/gpl-3.0.en.html).

**Author: Tim Wiese<br />
Affiliation: [European Space Agency](https://www.esa.int/)<br />
Maintainer: Tim Wiese, tim.wiese@esa.int**

This package has been tested under [ROS2] Foxy Fitzroy and Ubuntu 20.04. This is research code, expect that it changes often and any fitness for a particular purpose is disclaimed.

## Installation

### Building from Source

#### Dependencies

- [Robot Operating System (ROS)](http://wiki.ros.org) (middleware for robotics)
- [platform_driver_ethercat] (library for Elmo Gold Twitter and ATI FTS communication over EtherCAT using [SOEM]
- [rover_msgs](https://github.com/esa-prl/rover_msgs) (message definitions for ESA-PRL rovers)
- [yaml-cpp](https://github.com/jbeder/yaml-cpp) (a YAML parser and emitter in C++)

#### Building

To build from source, clone the latest version from this repository into your ros2 workspace and compile the package using

	cd ros2_ws/src
	git clone https://github.com/esa-prl/platform_driver_ethercat_ros2.git
	cd ../
	colcon build

> 

Reviewer 2: Please document here, that some error messages might appear and what to do if it takes ages to compile due to the password query.

Author: Not needed any more, since I commented out the `setcap` commands in CMakeLists for now: https://github.com/esa-prl/platform_driver_ethercat_ros2/pull/3/commits/e9e94cf1a7105c67d967b2ef1beca1860661d027

## Usage

Since the [platform_driver_ethercat] library uses [SOEM] which needs raw socket access, the node needs to be run with root privileges. This could be circumvented in the future with `setcap` and https://github.com/ros2/rcpputils/pull/44.

Change into a root shell while preserving your environment with

    sudo -sE

Then launch the main node with

	ros2 launch platform_driver_ethercat_ros2 platform_driver_ethercat.launch.py pd_config_file:=/path/to/config/file

## Launch files

launch/

* **platform_driver_ethercat.launch.py** Minimal launch file to start the node and bring it to the `Active` state. Needs to be executed with root privileges.

    Arguments:

    - **`pd_config_file`** Full path to the configuration file

## Config files

config/

> * **pd_marta.yaml** Configuration file for the MaRTA rover, to configure SOEM, Elmo drives and joints. This is NOT a regular ROS 2 parameter file. It makes use of advanced yaml features and is parsed inside the node.

Reviewer: ```suggestion
* **pd_marta.yaml** Configuration file for the MaRTA rover, to configure SOEM, Elmo drives and joints. This is NOT a regular ROS2 parameter file. It makes use of advanced yaml features and is parsed inside the node.
```

> * **pd_marta.yaml** Configuration file for the MaRTA rover, to configure SOEM, Elmo drives and joints. This is NOT a regular ROS 2 parameter file. It makes use of advanced yaml features and is parsed inside the node.

Reviewer 2: ```suggestion
* **pd_marta.yaml** Configuration file for the MaRTA rover, to configure SOEM, Elmo drives and joints. This is **NOT** a regular ROS 2 parameter file. It makes use of advanced yaml features and is parsed inside the node.
```

Author: Accepted

> * **pd_marta.yaml** Configuration file for the MaRTA rover, to configure SOEM, Elmo drives and joints. This is NOT a regular ROS 2 parameter file. It makes use of advanced yaml features and is parsed inside the node.

Reviewer 2: I suggest to either document here or in the config file what we'd have to change to make the config file work w/ ros parameters. It would also help to reference the `poses` parameters from locomotion_mode.

Author: I'd suggest to document that in a separate issue

Author: Added a comment to the config file and added additional info to the kanboard task

## Nodes

### platform_driver_ethercat_node

Lifecycle node to interface with the [platform_driver_ethercat] library.

#### Subscribed Topics

* **`joint_cmds`** ([rover_msgs/JointCommandArray](https://github.com/esa-prl/rover_msgs))

> 	Commanded position, velocity or torque for joints

Reviewer 2: ```suggestion
	Commanded position, velocity or torque for joints.
```

Author: Debatable as it is not a full sentence...

#### Published Topics

> * **`joint_states`** ([sensor_msgs/JointState](http://docs.ros.org/melodic/api/sensor_msgs/html/msg/JointState.html))

Reviewer 2: ```suggestion
* **`/joint_states`** ([sensor_msgs/JointState](http://docs.ros.org/melodic/api/sensor_msgs/html/msg/JointState.html))
```

Author: Wouldn't that mean that it's a global topic name? http://wiki.ros.org/Names

> 	Measured position, velocity and effort of joints

Reviewer 2: ```suggestion
	Measured position, velocity and effort of joints.
```

Author: Debatable as it is not a full sentence...

> * **`fts_readings`** ([rover_msgs/WrenchStampedArray](https://github.com/esa-prl/rover_msgs))

Reviewer 2: ```suggestion
* **`/fts_readings`** ([rover_msgs/WrenchStampedArray](https://github.com/esa-prl/rover_msgs))
```

Author: See above

> 	Measured force and torque values of sensors

Reviewer 2: ```suggestion
	Measured force and torque values of sensors.
```

Author: Debatable as it is not a full sentence...

> * **`temp_readings`** ([rover_msgs/TemperatureArray](https://github.com/esa-prl/rover_msgs))

Reviewer 2: ```suggestion
* **`/temp_readings`** ([rover_msgs/TemperatureArray](https://github.com/esa-prl/rover_msgs))
```

Author: See above

> 	Measured temperatures in degree Celsius

Reviewer 2: ```suggestion
	Measured temperatures in degree Celsius.
```

Author: See above

#### Parameters

* **`config_file`** (string)

> 	Full path to the configuration file

Reviewer 2: ```suggestion
	Full path to the configuration file.
```

Author: See above

## Bugs & Feature Requests

> Please report bugs and request features using the github issue tracker.

Reviewer 2: ```suggestion
Please report bugs and request features using the GitHub issue tracker.
```
Just noticed that github is also written wrong in ros2_best_practices. Will update template README.

Author: Accepted
